pkg/controller/pravegacluster/upgrade.go

```diff
@@ -111,10 +121,61 @@ func (r *ReconcilePravegaCluster) clearUpgradeStatus(p *pravegav1alpha1.PravegaC
 	return nil
 }
 
-func (r *ReconcilePravegaCluster) syncComponentsVersion(p *pravegav1alpha1.PravegaCluster) (err error) {
-	var synced bool
+func (r *ReconcilePravegaCluster) rollbackClusterVersion(p *pravegav1alpha1.PravegaCluster, version string) (err error) {
+	_, rollbackCondition := p.Status.GetClusterCondition(pravegav1alpha1.ClusterConditionRollback)
+	if rollbackCondition == nil || rollbackCondition.Status != corev1.ConditionTrue {
+		// We're in the first iteration for Rollback
+		// Add Rollback Condition to Cluster Status
+		log.Printf("Updating Target Version to  %v", version)
+		p.Status.TargetVersion = version
+		p.Status.SetRollbackConditionTrue("", "")
+		updateErr := r.client.Status().Update(context.TODO(), p)
+		if updateErr != nil {
+			p.Status.SetRollbackConditionFalse()
+			log.Printf("Error updating cluster: %v", updateErr.Error())
+			return fmt.Errorf("Error updating cluster status. %v", updateErr)
+		}
+	}
+
+	syncCompleted, err := r.syncComponentsVersion(p)
```

Comment:
For my understanding, does this call block until the update of all components complete?

Reply:
Nope. It updates the spec of one component and then on every reconciliation loop checks how many replicas are updated for that component. When all are updated, we move to next component till all components are upgraded.